frontend/expr_inferrer.py

```diff
@@ -124,6 +128,14 @@ def _infer_add(left_type, right_type, lineno, solver):
     result_type = solver.new_z3_const("addition_result")
     solver.add(axioms.add(left_type, right_type, result_type, solver.z3_types),
                fail_message="Addition in line {}".format(lineno))
+
+    # Add soft constraint in case of list addition
```

Comment:
I don't think we need soft constraints here. Because the type of the addition result (and similarly the other operations) is assigned directly (with equality not subtype relationship) to the type of the stronger of the two operands.